Framework/PythonInterface/mantid/FunctionWrapper.py

```diff
@@ -0,0 +1,392 @@
+             currentFunction = self[i].flatten()
+             for j in range(0, len(currentFunction)):
+                flatSelf.fun.add(currentFunction[j].fun)
+          else:
+             flatSelf.fun.add(self[i].fun)
+             
+       return flatSelf          
+ 
+class ProductFunctionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Product Fitting Function
+    """
+    def __init__ (self, *args):
+       """ Called when creating an instance
+           It should not be called directly.
+       """
+       return self.initByName("ProductFunction", *args)
+       
+class ConvolutionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Convolution Fitting Function
+    """
+    def __init__ (self, *args):
+       """ Called when creating an instance
+           It should not be called directly.
+       """
+       return self.initByName("Convolution", *args)
+       
+class MultiDomainFunctionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Product Fitting Function
+    """
+    def __init__ (self, *args, **kwargs):
+       """ Called when creating an instance
+           It should not be called directly
+       """
+       # Create and populate with copied functions
+       self.initByName("MultiDomainFunction", *args)
+               
+       # Tie the global parameters
+       if 'Global' in kwargs:
+          list = kwargs['Global']
+          for name in list:
+             self.tieAll(name)
+       
+       # Set domain indices: 1 to 1       
+       for i in range(0, len(self)):
+          self.fun.setDomainIndex(i, i)
+          
+    def nDomains (self):
+       """ Return number of domains 
+           Used for testing
+       """
+       return self.fun.nDomains()
+     
+        
+def _create_wrapper_function(name):
+    """Create fake functions for the given name
+       It should not be called directly
+                  
+       :param name: name of fake function
+    """
+    # ------------------------------------------------------------------------------------------------
+    def wrapper_function(*args, **kwargs):
+        if  name == "CompositeFunction":
+           return CompositeFunctionWrapper( *args, **kwargs )
+        elif name == "ProductFunction":
+           return ProductFunctionWrapper( *args, **kwargs )
+        elif name == "Convolution":
+           return ConvolutionWrapper( *args, **kwargs )
+        elif name == "MultiDomainFunction":
+           return MultiDomainFunctionWrapper( *args, **kwargs )
+        else:
+           return FunctionWrapper(name, *args, **kwargs)
+
+    # ------------------------------------------------------------------------------------------------
+    wrapper_function.__name__ = name
+    # _replace_signature(fake_function, ("", ""))
+    globals()[name] = wrapper_function
+
+fnames = FunctionFactory.getFunctionNames()
+for i, val in enumerate(fnames):    
+   _create_wrapper_function(val)
+
```

Comment:
Some of the specifications have not been implemented.
Run the test scripts below to see what is missing.

1.
```
p = Polynomial(n=3, A0=1, A1=2, A2=4, A3=3)
print p
p.n = 4
p.A0 = 1.1
print p
print p['n']
print p.n
print p.A0
```
2.
```
g1 = Gaussian(Sigma=.5,Height=3)
g2 = Gaussian(PeakCentre=3, Sigma=.7, Height=4)
b = LinearBackground(A0=1, A1=0.2)
f = b + g1 + g2
print len(f) # must be 3
print f[0]
print f[1]

f = b * g1 * g2
print len(f) # must be 3
print f[0]
print f[1]

```
3.
```
g = Gaussian(Sigma=.5,Height=3)
b = LinearBackground(A0=1, A1=0.2)
f = b + g
print f
print f.f0
print f.f0.A0
```
4.
```
g = Gaussian(Sigma=.5,Height=3)
b = LinearBackground(A0=1, A1=0.2)
f = b + g + g

print f.LinearBackground
print f.Gaussian0
print f.Gaussian1
```
5.
```
g = Gaussian(Sigma=.5,Height=3)
b = LinearBackground(A0=1, A1=0.2)
f = b + g
f[1] = Lorentzian(PeakCentre=2)
```
6.
```
p = Polynomial(n=3, A0=1, A1=2, A2=4, A3=3)
p.tie(A0=2)
p.tie(A1='A2+A3')
p.tie({'A2': '2*A3'})
print p
f = (Gaussian() + Gaussian() + Gaussian()).flatten()
f.tie({'f1.Sigma': 'f0.Sigma', 'f2.Sigma': 'f0.Sigma'})
print f
f = (Gaussian() + Gaussian() + Gaussian()).flatten()
f.tieAll('Sigma')
print f
f.untie('f2.Sigma')
print f
f.untie('f0.Sigma')
print f
f.untie('f1.Sigma')
print f
f.tieAll('Sigma')
print f
f.untie('f1.Sigma', 'f2.Sigma')
print f
f.tieAll('Sigma')
f.untieAll()
print f
```
7.
```
p = Polynomial(n=3, A0=1, A1=2, A2=4, A3=3)
p.fix('A0')
p.fix('A0', 'A2')
print p
p.unfix('A0')
p.unfix('A0', 'A2')
print p
p.unfixAll()
print p
p.fixAll()
```
8.
```
g = Gaussian()
g.constrain('-1 < PeakCentre < 1', 'Sigma > 0')
f = (g + g + g).flatten()
f.constrainAll('Sigma>0')
f.constrainAll('Height > 1', 'Sigma > 0')
print f
```